### ISSUE TITLE
pdHess=TRUE while model failed to converge

### ISSUE THREAD

**Issue by KarinePrince** [glmmTMB/glmmTMB#275]
Hi,

I apologize for re-opening a discussion that has already been closed (about "Convergence diagnostics"), but I am currently trying to fit a Poisson model to count time series data, and I am wondering about the accuracy of *sdr$pdHess=TRUE*.
When determining whether my model has converged though, I found myself in trouble trying to pick up some convergence criteria... 
*pdHess* does not seem to be robust for that (see below).
Instead, I found that testing *if(identical(is.na(mymodel$sdr$diag.cov.random)[1], FALSE)* seems to provide the good information about whether the model has converged. Does that sound correct?
(we agree on the fact that, parameter estimates without std errors or pvalues reveal a problem of model convergence?...)

Here are examples of what I could get after running my model (model tested on different species).
**Situation 1:**
- no warning message about model convergence
- sdr$pdHess = TRUE
- estimates of parameters but no std errors and pvalues estimated

**Situation 2:**
- warning message about model convergence
- sdr$pdHess = TRUE
- estimates of parameters but no std errors and pvalues estimated

Here is an example of the outputs I got.
> mymodel <- glmmTMB( count ~ scale(year) + habitat + scale(ratio_EPfaibles) + scale(ratio_EPfortes) + long_wgs84 + lat_wgs84 + (1|site), data = data_sp, family = poisson )
Warning message:
In fitTMB(TMBStruc) :
  Model convergence problem; extreme or very small eigen values detected. See vignette('troubleshooting')

> summary(mymodel)
Family: poisson  ( log )
Formula:          count ~ scale(year) + habitat + scale(ratio_EPfaibles) + scale(ratio_EPfortes) +      long_wgs84 + lat_wgs84 + (1 | site)
Data: data_sp

     AIC      BIC   logLik deviance df.resid 
   332.7    397.0   -154.3    308.7     1565 

Random effects:

Conditional model:
 Groups Name        Variance Std.Dev.
 site   (Intercept) 8.809    2.968   
Number of obs: 1577, groups:  site, 231

Conditional model:
                        Estimate Std. Error z value Pr(>|z|)
(Intercept)            -61.08098         NA      NA       NA
scale(year)             -0.04989         NA      NA       NA
habitatmixed             5.01333         NA      NA       NA
habitatother           -43.12072         NA      NA       NA
habitaturban             3.41056         NA      NA       NA
habitatwetland          -0.13095         NA      NA       NA
habitatwoodland         -1.79499         NA      NA       NA
scale(ratio_EPfaibles)   0.52218         NA      NA       NA
scale(ratio_EPfortes)    0.88057         NA      NA       NA
long_wgs84               0.97146         NA      NA       NA
lat_wgs84                1.06925         NA      NA       NA

**Comment by bbolker** [comment]
Yes, getting `NA` values for the standard errors suggests some kind of convergence issue. A reproducible example would be fantastic, if you can provide one ...

**Comment by KarinePrince** [comment]
[ACRRIS_data.txt](https://github.com/glmmTMB/glmmTMB/files/1491908/ACRRIS_data.txt)
Thanks for taking the time to reply!

Attached is a .txt with the required data. For that particular species the Poisson model won't converge, but nbinom2 distribution seems to be suitable.
And here is a subset (simplified version) of the code I run:

data_sp <- read.table("ACRRIS_data.txt", h=T)
## run glmm - fit poisson model with site as random
glmmPois.sp <-
  glmmTMB(
    count ~ scale(year) + habitat + scale(ratio_EPfaibles) + scale(ratio_EPfortes) + long_wgs84 + lat_wgs84 + (1 |site),
    data = data_sp,
    family = poisson
  )

## test overdispersion
obs <- as.numeric(data_sp$count)
resid <- residuals(glmmPois.sp, type = c("response"))
pred <- (obs - resid)
z <- (((obs - pred) ^ 2) - obs) / (pred * sqrt(2))
res <- summary(lm(z ~ 1))
test <- res$coefficients[4] < 0.05
print(test)


## if overdispersion TRUE, fit a negative binomial distribution with zi parameter
## also, test for model convergence and try to fit other models if convergence is failing
if (isTRUE(test)) {
  #here, if overdispersion=T, run NB mdl
  
  #try a standard negative binomial model - NB1 (variance = φμ). NB1 similar to quasi-poisson
  try(glmmNB1.sp <-
        update(glmmPois.sp, family = nbinom1),
      silent = T)
  
  #try a standard negative binomial model - NB1 (variance = φμ), with zero-inflation
  try(glmmZiNB1.sp <-
        update(glmmPois.sp, ziformula =  ~ 1, family = nbinom1),
      silent = T)
  
  if (isTRUE(glmmNB1.sp$sdr$pdHess) &
      AIC(glmmNB1.sp) < AIC(glmmZiNB1.sp)) {
    #here: if nbinom1 converge and fit of glmmNB1 is better than glmmZiNB1, apply nbinom1 mdl
    model.sp <- glmmNB1.sp
    mdl.mess <-
      as.vector("outputs from glmmZiNB1") #message indicating which model was chosen. further saved as output
  } else if (isTRUE(glmmZiNB1.sp$sdr$pdHess)) {
    #here: if nbinom1 did not converge but zi nbinom1 did converge, apply zi nbinom1 mdl
    model.sp <- glmmZiNB1.sp
    mdl.mess <- as.vector("outputs from glmmZiNB1")
  } else {
    #here: if any model converge keep poisson and save error message
    model.sp <- glmmPois.sp
    mdl.mess <-
      as.vector("outputs from model.sp / no suitable model")
  }
} else if (identical(isTRUE(glmmPois.sp$sdr$pdHess), FALSE)) {
  #here: if overdispersion=F, but Poisson mdl did not converge, try to fit another distribution
  
  #try a standard negative binomial model - NB2 (variance = μ(1+μ/k))
  try(glmmNB2.sp <-
        update(glmmPois.sp, family = nbinom2),
      silent = T)
  
  model.sp <- glmmNB2.sp
  mdl.mess <- as.vector("outputs from glmmNB2")
} else {
  #here, if overdispersion=F and poisson mdl converged, keep poisson mdl
  model.sp <- glmmPois.sp
  mdl.mess <- as.vector("outputs from glmmPois")
}

model.sp <<- model.sp
mdl.mess <<- mdl.mess
